### DISCUSSION TITLE
The right way to make all Firestore requests run on the server side

### DISCUSSION THREAD

**Discussion by alihardan** [quasarframework/quasar#15740] [General - Components / Directives / etc] [answered]
Hello everyone, I have little experience in working with the Quasar framework. Documentation is excellent but some things about SSR are still unclear to me. :cry: 

I'm building a CRUD with the power of Quasar, and using Firestore as a database.  I have enabled SSR mode. I want all HTTP requests to go through my server and users should never connect directly to Firestore. (Because Firebase has banned access from some regions and countries).

I can fetch data on the server side by using the PreFetch feature. but I wonder how to manage after rendering actions such as updating the Firestore database on-fly. Should I make custom APIs on the Express.js server for these requests? Or is there a better way that exists?

Thanks.

**Comment by TobyMosque** [comment]
yes, but in at your place i would configure the api inside a ssr middleware.
https://quasar.dev/quasar-cli-webpack/developing-ssr/ssr-middleware/

**Reply by TobyMosque** [reply, marked as answer]
just curious, do u tried to put the firebase behind your api using a reverse proxy?

something like:
https://github.com/chimurai/http-proxy-middleware
```js
import { createProxyMiddleware } from 'http-proxy-middleware';
import { ssrMiddleware } from 'quasar/wrappers'

export default ssrMiddleware(({ app }) => {
  app.use(
    '/firebase',
    createProxyMiddleware({
      target: 'https://xxx.firebaseio.com/',
      changeOrigin: true,
      pathRewrite: { '^/firebase': '' },
      ws: true
    })
  );
})
```

so, in your app.:
```ts
import { boot } from 'quasar/wrappers';
import { initializeApp } from 'firebase/app';
import { fbAppKey } from 'src/composables/firebase';

export default boot(async ({ app, store }) => {
  const dbUrl = process.env.SERVER
    ? "https://xxx.firebaseio.com"
    : "/firebase";
  const fbApp = initializeApp({
    "databaseURL": dbUrl,
    /* your other options */
  });

  app.provide(fbAppKey, fbApp);
});
```

**Reply by alihardan** [reply]
WoW! It's the best idea. Sadly didn't succeed to customize the Firestore domain. the `databaseURL` is for Realtime Database only. :disappointed:

**Reply by alihardan** [reply]
Fortunately it's not hardcoded, I succeeded to change the Firestore host:

``` js
const db = initializeFirestore(app, { host: "example.com" });
```

I will test the reverse proxy and I will tell you the result.

**Reply by alihardan** [reply]
@TobyMosque, you are a genius. 🤩👏🏻 It is hard to believe that it was solved so easily. Here is what I wrote in my app:

``` js
  const app = initializeApp(firebaseConfig);
  const db = initializeFirestore(app, {
    host: "localhost:9100", // Set your domain here
    ssl: false, // I disabled SSL because I was on development environment
  });
```

and my `src-ssr/middlewares/render.js`:
``` js
import { createProxyMiddleware } from "http-proxy-middleware";
import { ssrMiddleware } from "quasar/wrappers";

export default ssrMiddleware(({ app, resolve, render, serve }) => {
  app.use(
    "/google.firestore.v1.Firestore",
    createProxyMiddleware({
      target: "https://firestore.googleapis.com/google.firestore.v1.Firestore/",
      changeOrigin: true,
      pathRewrite: { "^/google.firestore.v1.Firestore": "" },
      ws: true,
    })
  );

  /*
    ...
  */
})
```

If there are any suggestions or better practices I'd be happy to see them here.
